src/server/tran_server.cpp

```diff
@@ -362,75 +370,56 @@ void
-  conn_vec.clear ();
-  m_main_conn_cv.notify_all ();
-
-  // Wait until all disconnection reuqests are handled.
-  m_async_disconnect_handler.terminate ();
+  for (auto &conn : m_page_server_conn_vec)
+    {
+      constexpr bool with_disconnect_msg = true;
+      conn->disconnect_async (with_disconnect_msg);
+    }
+  m_page_server_conn_vec.clear ();
 
   er_log_debug (ARG_FILE_LINE, "Transaction server disconnected from all page servers.");
 }
 
-void
-tran_server::disconnect_page_server_async (const connection_handler *conn)
+int
+tran_server::reset_main_connection ()
 {
-  bool main_conn_changed = false;
-  std::string prev_main_conn_id;
   auto &conn_vec = m_page_server_conn_vec;
-  auto ulock = std::unique_lock<std::shared_mutex> (m_page_server_conn_vec_mtx);
-
-  auto conn_it = std::find_if (conn_vec.begin (), conn_vec.end (),
-			       [&conn] (std::unique_ptr<connection_handler> &conn_uptr)
+  /* the priority to select the main connection is the order in the container */
+  const auto main_conn_cand = std::find_if (conn_vec.cbegin (), conn_vec.cend (),
+			      [] (const auto &conn)
   {
-    return conn_uptr.get () == conn;
+    return conn->is_connected ();
   });
 
-  if (conn_it == conn_vec.end ())
+  if (main_conn_cand == conn_vec.cend())
     {
-      return; // the connection is already cleared in disconnect_all_page_servers()
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CONN_NO_PAGE_SERVER_AVAILABLE, 0);
+      return ER_CONN_NO_PAGE_SERVER_AVAILABLE;
     }
 
-  if (conn_it == conn_vec.begin ())
+  auto ulock = std::unique_lock<std::shared_mutex> { m_main_conn_mtx };
+  if (m_main_conn == main_conn_cand->get ())
     {
-      main_conn_changed = true;
-      prev_main_conn_id = (*conn_it)->get_channel_id ();
+      er_log_debug (ARG_FILE_LINE, "The main connection has been already reset.\n");
     }
-
-  m_async_disconnect_handler.disconnect (std::move (*conn_it));
-  assert (*conn_it == nullptr);
-  conn_vec.erase (conn_it);
-
-  if (main_conn_changed)
+  else
     {
-      if (conn_vec.empty ())
-	{
-	  er_log_debug (ARG_FILE_LINE, "The last connection is disconnected (%s). No main connection available.\n",
-			prev_main_conn_id.c_str ());
-	}
-      else
-	{
-	  er_log_debug (ARG_FILE_LINE, "The main connection is changed from %s to %s.\n",
-			prev_main_conn_id.c_str (), (*conn_vec.begin ())->get_channel_id ().c_str ());
-	}
-      ulock.unlock ();
-      m_main_conn_cv.notify_all ();
+      m_main_conn = main_conn_cand->get ();
+      er_log_debug (ARG_FILE_LINE, "The main connection is set to %s.\n",
+		    m_main_conn->get_channel_id ().c_str ());
     }
+
+  return NO_ERROR;
 }
 
 bool
 tran_server::is_page_server_connected () const
 {
   assert_is_tran_server ();
-
-  return !m_page_server_conn_vec.empty ();
+  return std::any_of (m_page_server_conn_vec.cbegin (), m_page_server_conn_vec.cend (),	[] (const auto &conn)
+  {
+    return conn->is_connected ();
+  });
```

Comment:
```suggestion
  return std::any_of (m_page_server_conn_vec.cbegin (), m_page_server_conn_vec.cend (),	[] (const auto &conn)
  {
    return conn->is_connected ();
  }) && (m_main_conn != nullptr);
```
or at least an assert